In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt

# Задание 1

Найдите минимально необходимый объем выборки для построения интервальной оценки среднего с точностью ∆ = 3, дисперсией σ^2 = 225 и уровнем доверия β = 0.95.

In [2]:
sigma = 15
Z = 1.96  # 95%
e = 3
N = ((sigma*Z)/e)**2

print('Ответ: минимально необходимый объем выборки -',round(N,2))

Ответ: минимально необходимый объем выборки - 96.04


# Задание 2

Вам даны две выборки роста мужчин и женщин. Докажите, используя t-Тест Стьюдента, что различия между выборками незначительны, если уровень значимости равен 0.001

In [4]:
np.random.seed(123)
population_men=stats.norm.rvs(loc=19,scale=171,size=11000000)  # Выборка мужчин со средним ростом 171
population_women=stats.norm.rvs(loc=16,scale=165,size=12000)   # Выборка женщин со средним ростом 165

In [5]:
from math import sqrt
from numpy.random import seed
from numpy.random import randn
from numpy import mean
from scipy.stats import sem
from scipy.stats import t
 
# function for calculating the t-test for two independent samples
def independent_ttest(data1, data2, alpha):
	# calculate means
	mean1, mean2 = mean(data1), mean(data2)
	# calculate standard errors
	se1, se2 = sem(data1), sem(data2)
	# standard error on the difference between the samples
	sed = sqrt(se1**2.0 + se2**2.0)
	# calculate the t statistic
	t_stat = (mean1 - mean2) / sed
	# degrees of freedom
	df = len(data1) + len(data2) - 2
	# calculate the critical value
	cv = t.ppf(1.0 - alpha, df)
	# calculate the p-value
	p = (1.0 - t.cdf(abs(t_stat), df)) * 2.0
	# return everything
	return t_stat, df, cv, p
 
# calculate the t test
alpha = 0.001
t_stat, df, cv, p = independent_ttest(population_men, population_women, alpha)
print('t=%.3f, df=%d, cv=%.3f, p=%.3f' % (t_stat, df, cv, p))
# interpret via critical value
if abs(t_stat) <= cv:
	print('Accept null hypothesis that the means are equal.')
else:
	print('Reject the null hypothesis that the means are equal.')
# interpret via p-value
if p > alpha:
	print('Accept null hypothesis that the means are equal.')
else:
	print('Reject the null hypothesis that the means are equal.')

t=0.195, df=11011998, cv=3.090, p=0.845
Accept null hypothesis that the means are equal.
Accept null hypothesis that the means are equal.


Ответ: мы обнаруживаем, что имеется достаточно доказательств, чтобы принять нулевую гипотезу, обнаружив, что выборочные средние, вероятно, равны. Т.е. различия между выборками незначительны при уровне значимости 0.001

# Задание 3

Определите объем необходимой выборки для исследования среднего чека за кофе в случайном городе, если известно, что в этом городе стандартное отклонение = 150, уровень доверия = 95%. Погрешность 50 рублей.

In [17]:
sigma = 150
Z = 1.96  # 95%
e = 50
N = ((sigma*Z)/e)**2

print('Ответ: объем необходимой выборки -',round(N))

Ответ: объем необходимой выборки - 35


# Задание 4

Представьте, что вы хотите разоблачить "волшебника", который считает, что умеет предсказывать погоду на завтра. Отвечая просто: дождь или солнце. Вы пронаблюдали за ответами "волшебника" в течение какого-то периода времени и получили такие результаты (см.ниже). Можно ли сказать, что маг действительно умеет предсказывать погоду, если уровень значимости принять за 0.05?

In [7]:
observations = pd.DataFrame([[25,36],[15,44]],
                            index=['Дождь','Солнце'],
                            columns=['Ответ волшебника','Реальность'])
observations

Ответ волшебника  Реальность
Дождь                 25          36
Солнце                15          44

In [23]:
# Нулевая гипотеза состоит в том, что все совпадения случайны и "волшебник" не умеет предсказывать погоду

import scipy as sp
oddsratio, pvalue = sp.stats.fisher_exact(observations)
print(oddsratio, pvalue)

2.037037037037037 0.08323639416753109


Ответ: в нашем примере р > 0,05, значит, принимается нулевая гипотеза и делается вывод, что маг не умеет предсказывать погоду.

# Задание 5

Используя функцию mean_confidence_interval(data, confidence), постройте доверительный интервал с уровнем доверия 90% для выборки init_data 

In [9]:
init_data = [4,5,8,9,6,7,5,6,7,8,5,6,7,0,9,8,4,6,7,9,8,6,5,7,8,9,6,7,5,8,6,7,9,5,10]

In [24]:
import scipy.stats

def mean_confidence_interval(data, confidence):
    n = len(data)
    m, se = np.mean(data), scipy.stats.sem(data)
    h = se * scipy.stats.t.ppf((1 + confidence)/2, n)
    return m-h, m+h


print('Ответ: ', mean_confidence_interval(init_data, 0.90))

Ответ:  (6.074074746111319, 7.183068111031539)


# Задание 6

Принадлежит ли выборка data_1 и data_2 одному множеству? Оцените это с помощью известных вам тестов проверки гипотез.

In [25]:
data_1 = [4,5,8,9,6,7,5,6,7,8,5,6,7,0,9,8,4,6,7,9,8,6,5,7,8,9,6,7,5,8,6,7,9,5,10]
data_2 = [8,5,6,7,0,1,8,4,6,7,0,2,6,5,7,5,3,5,3,5,3,5,5,8,7,6,4,5,3,5,4,6,4,5,3,2,6,4,2,6,1,0,4,3,5,4,3,4,5,4,3,4,5,4,3,4,5,3,4,4,1,2,4,3,1,2,4,3,2,1,5,3,4,6,4,5,3,2,4,5,6,4,3,1,3,5,3,4,4,4,2,5,3]

# Формулируем нулевую гипотезу: данные выборки принадлежат одному множеству

In [32]:
# T-test Стьюдента
from scipy import stats

t , p  = stats.ttest_ind(data_1,data_2)
print("t = " + str(t))
print("p = " + str(p))
if p < 0.05:
    print("reject null hypothesis")
else:
    print("accept null hypothesis")

t = 7.35744682757447
p = 2.1159475572725764e-11
reject null hypothesis


In [31]:
# One Way F-test(Anova)
F, p = stats.f_oneway(data_1, data_2)
print("p-value for significance is: ", p)
if p < 0.05:
    print("reject null hypothesis")
else:
    print("accept null hypothesis")

p-value for significance is:  2.115947557272575e-11
reject null hypothesis


Ответ: данные тесты показали, что есть все основания отвергнуть нулевую гипотезу и сделать вывод, что выборки не относятся к одному множеству.

# Задание 7

На примере датасета про жилье в New York City, мы сталкивались с примером, когда переменная имеет не совсем нормальное распределение.
Предположим, Вы сформировали две гипотезы: Нулевая гипотеза - распределение нормальное, Альтернативная гипотеза - распределение не нормальное.
Допустим, вы применили какой-то тест (сейчас неважно какой), который показал уровень значимости (p-value) = 0.03. Каковы будут ваши выводы? Будем считать что у нас нормальное распределение или все-таки нет? Вопрос без подвоха)

Ответ: если мы заранее определили уровень статистической значимости - 0,05, то при р < 0,05 делаем вывод, что аргументов достаточно, чтобы отвергнуть нулевую гипотезу, хотя есть небольшой шанс против этого. То есть будем считать, что распределение не нормальное.

# Задание 8

Первая выборка — это пациенты, которых лечили препаратом А. Вторая выборка — пациенты, которых лечили препаратом Б. Значения в выборках — это некоторая характеристика эффективности лечения (уровень метаболита в крови, температура через три дня после начала лечения, срок выздоровления, число койко-дней, и т.д.)

а) Требуется выяснить, имеется ли значимое различие эффективности препаратов А и Б, или различия являются чисто случайными и объясняются «естественной» дисперсией выбранной характеристики? (уровень значимости принять за 5% или 0.05)

b) При каком минимальном P-values различия были бы уже значимы?

In [35]:
# T-тест с двумя выборками всегда использует следующую нулевую гипотезу: различия являются случайными

np.random.seed(11)
A = stats.norm.rvs(scale=50,loc=10,size=300)
B = A+stats.norm.rvs(scale=10,loc=-1.25,size=300)

t, p = stats.ttest_ind(A, B, equal_var=False)
print("t = " + str(t))
print("p = " + str(p))

t = 0.33196234898293125
p = 0.7400341626811637


а) Ответ: поскольку р-значение равно 0,74, то есть больше 0,05 (или 5%), мы не можем отвергнуть нулевую гипотезу. Можно сделать вывод, что значимого различия эффективности препаратов А и Б нет. Сказать, что различие есть, значит рискнуть ошибиться на 74%.

b) Уровень значимости принимается исследователем и зависит от поставленных им целей. Но в науке принят компромисс: у нас есть основания отвергнуть нулевую гипотезу при  P-values < 0.05.